shinylive/_url.py

```diff
@@ -0,0 +1,187 @@
+from __future__ import annotations
+
+import base64
+import json
+import os
+from pathlib import Path
+from typing import List, Literal, Optional, TypedDict, cast
+
+
+class FileContentJson(TypedDict):
+    name: str
+    content: str
+    type: Literal["text", "binary"]
+
+
+def encode_shinylive_url(
+    app: str | Path,
+    files: Optional[tuple[str | Path, ...]] = None,
+    mode: str = "editor",
+    language: Optional[str] = None,
+    header: bool = True,
+) -> str:
+    """
+    Generate a URL for a [ShinyLive application](https://shinylive.io).
+
+    Parameters
+    ----------
+    app
+        The main app file of the ShinyLive application. This file should be a Python
+        `app.py` or an R `app.R`, `ui.R`, or `server.R` file. This file will be renamed
+         `app.py` or `app.R` for shinylive, unless it's named `ui.R` or `server.R`.
+    files
+        A tuple of file or directory paths to include in the application. On shinylive,
+        these files will be stored relative to the main `app` file. If an entry in files
+        is a directory, then all files in that directory will be included, recursively.
+    mode
+        The mode of the application. Defaults to "editor".
+    language
+        The language of the application. Defaults to None.
+    header
+        Whether to include a header. Defaults to True.
+
+    Returns
+    -------
+        The generated URL for the ShinyLive application.
+    """
+
+    # if app has a newline, then it's app content, not a path
+    if isinstance(app, str) and "\n" in app:
+        # now language is required
+        if language is None:
+            raise ValueError("If `app` is a string, then `language` must be specified.")
```

Comment:
We could also inspect `app` for `import shiny`, `from shiny` or `library(shiny)` to decide, before bailing.